GNUmakefile: add spellfix target, use it.

dkegel-fastly opened the pull request:
Also:
- point to an updated version of the misspell package.
- give internal/tools its own go.mod temporarily because new misspell needs a newer go than toplevel go.mod specifies
- make lint target depend on tools target and use the revive it installs (assuming go install causes it to be found via PATH)
- fix what newer misspell and lint find
- add spell to CI

dkegel-fastly committed:
builder/cc1as.h: fix typo found by 'make spell'

dkegel-fastly committed:
GNUMakefile: remove exception for inbetween, fix instance now found by 'make spell'

dkegel-fastly committed:
GNUMakefile: remove exception for programmmer, fix instance now found by 'make spell'

dkegel-fastly committed:
go.mod: use updated misspell. GNUmakefile: add spellfix target, use it.

Also ignore directories properly when invoking spellchecker.

Also run go mod tidy (with go 1.19)

dgryski commented on compiler/compiler.go:
```diff
@@ -242,7 +242,7 @@ func NewTargetMachine(config *Config) (llvm.TargetMachine, error) {
 
 // Sizes returns a types.Sizes appropriate for the given target machine. It
-// includes the correct int size and aligment as is necessary for the Go
+// includes the correct int size and allignment as is necessary for the Go
```
`alignment`

dkegel-fastly replied:
unbroke

dgryski commented on src/internal/wasi/clocks/v0.2.0/monotonic-clock/monotonic-clock.wit.go:
```diff
@@ -78,7 +78,7 @@ func wasmimport_Resolution() (result0 uint64)
 //
 // Create a `pollable` which will resolve once the specified instant
-// occured.
+// occurred.
```
Sadly this is a misspelling upstream, *and* this is a generated file.  We should probably ignore `src/internal/wasi` for now.

dkegel-fastly replied:
ignored

dgryski commented on src/internal/wasi/io/v0.2.0/streams/streams.wit.go:
```diff
@@ -413,7 +413,7 @@ func wasmimport_OutputStreamFlush(self0 uint32, result *cm.Result[StreamError, s
 // Read from one stream and write to another.
 //
-// The behavior of splice is equivelant to:
+// The behavior of splice is equivalent to:
```
I might submit spelling fixes for the original `.wit` files upstream.

dgryski reviewed (changes requested):
I'm a fan of automation.  Even better if the tooling is reliable enough to run as part of CI.

dkegel-fastly commented:
Thanks for the review!  Ready for another look.

dgryski reviewed (approved):
LGTM

b0ch3nski commented on go.mod:
```diff
@@ -1,14 +1,16 @@
 module github.com/tinygo-org/tinygo
 
-go 1.19
+go 1.21
```
I think I saw somewhere on tinygo.org docs mentioning a requirement of Go 1.20 - would be good to confirm that and also bump the version in docs

dkegel-fastly replied:
Not yet!  I think we still work with go 1.19.  So the proper workaround is probably a separate go.mod just for the spell checker for the moment.

archie2x reviewed (approved):
LGTM

leongross commented on go.mod:
Ci [fails](https://github.com/tinygo-org/tinygo/actions/runs/10288011952/job/28472435670?pr=4387#step:19:39) with
```
go: updates to go.mod needed; to update it:
	go mod tidy
```
So update go modules and/or rebase against dev?

dkegel-fastly replied:
I suspect separate go.mod to specify a newer go for now, to be removed when top level go.mod catches up

dkegel-fastly committed:
make spell: give internal/tools its own go.mod, as misspell requires newer go

Remove the go.mod/go.sum in internal/tools once doing so doesn't break CI (e.g. once we drop support for go 1.19)

Also had to make the lint target depend on tools and run the installed revive
(which was perhaps implied by the change that added revive to internal/tools,
but not required in GNUmakefile until we gave internal/tools its own temporary go.mod)

Ditto for revive, and fix problems it found

dkegel-fastly committed:
.github: now that 'make spell' works well, run it from CI

Also cause make spell to abort if it finds misspelt words, so what it finds doesn't get lost in CI logs

dkegel-fastly committed:
GNUmakefile: tools: avoid -C option on go generate to make test-llvm15-go119 circleci job happy

Avoids error

    go generate -C ./internal/tools -tags tools  ./
    flag provided but not defined: -C
    usage: go generate [-run regexp] [-n] [-v] [-x] [build flags] [file.go... | packages]
    Run 'go help generate' for details.
    make: *** [GNUmakefile:952: tools] Error 2

The -C flag was added sometime around go 1.20 by
https://cs.opensource.google/go/go/+/2af48cbb7d85e5fdc635e75b99f949010c607786
so we can't use it while we still support go 1.19 in CI

dkegel-fastly committed:
internal/tools/go.mod: fix format of go version

Avoids error

    cd internal/tools && go generate -tags tools ./
    go: errors parsing go.mod:
    /root/project/internal/tools/go.mod:6: invalid go version '1.21.13': must match format 1.23
    make: *** [GNUmakefile:952: tools] Error 1

dgryski reviewed (approved):
LGTM
